glusterfsd/src/glusterfsd-mgmt.c

```diff
@@ -355,41 +356,20 @@ glusterfs_handle_terminate(rpcsvc_request_t *req)
     }
 
     glusterfs_terminate_response_send(req, 0);
-    for (trav_p = &top->children; *trav_p; trav_p = &(*trav_p)->next) {
-        tvictim = (*trav_p)->xlator;
-        if (!tvictim->cleanup_starting &&
-            !strcmp(tvictim->name, xlator_req.name)) {
-            continue;
-        }
-        if (!tvictim->cleanup_starting) {
-            still_bricks_attached = _gf_true;
-            break;
-        }
-    }
-    if (!still_bricks_attached) {
-        gf_log(THIS->name, GF_LOG_INFO,
-               "terminating after loss of last child %s", xlator_req.name);
-        rpc_clnt_mgmt_pmap_signout(glusterfsd_ctx, xlator_req.name);
-        kill(getpid(), SIGTERM);
-    } else {
-        /* TODO cleanup sequence needs to be done properly for
-           Quota and Changelog
-        */
-        if (victim->cleanup_starting)
-            goto err;
-
-        rpc_clnt_mgmt_pmap_signout(glusterfsd_ctx, xlator_req.name);
-        victim->cleanup_starting = 1;
+    if (victim->cleanup_starting)
+        goto err;
 
-        UNLOCK(&ctx->volfile_lock);
-        lockflag = _gf_true;
+    rpc_clnt_mgmt_pmap_signout(glusterfsd_ctx, xlator_req.name);
+    victim->cleanup_starting = 1;
+    if (totchild == 1)
```

Comment:
I'm not sure if this process is atomic enough to correctly detect the last stop of last child. For example, if there are two childs, one of them already stopping, and we want to stop the second one, `totchild` will be 2, but this is the last brick and `ctx->cleanup_starting` should be set.

How is this handled ?

Reply:
i have updated condition, totchild needs to be update only if cleanup_starting is not set.